modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java

```diff
@@ -280,38 +305,65 @@ private void throwIfChainBelongsToAnotherTx(VersionChain versionChain, UUID txId
         assert transactionId != null ^ timestamp != null;
 
         if (transactionId != null) {
-            return findLatestRowVersion(versionChain, transactionId, keyFilter);
+            ReadResult res = findLatestRowVersion(versionChain, transactionId, keyFilter);
+
+            if (res == null) {
+                return null;
+            }
+
+            return res.binaryRow();
         } else {
-            ByteBufferRow row = findRowVersionByTimestamp(versionChain, timestamp);
+            ReadResult res = findRowVersionByTimestamp(versionChain, timestamp);
+
+            if (res == null) {
+                return null;
+            }
+
+            BinaryRow row = res.binaryRow();
 
             return keyFilter.test(row) ? row : null;
         }
     }
 
-    private @Nullable ByteBufferRow findRowVersionByTimestamp(VersionChain versionChain, HybridTimestamp timestamp) {
-        if (!versionChain.hasCommittedVersions()) {
-            return null;
-        }
-
-        long newestCommittedLink = versionChain.newestCommittedLink();
+    private @Nullable ReadResult findRowVersionByTimestamp(VersionChain versionChain, HybridTimestamp timestamp) {
+        long headLink = versionChain.headLink();
 
         ScanVersionChainByTimestamp scanByTimestamp = new ScanVersionChainByTimestamp(partitionId);
 
         try {
-            rowVersionDataPageReader.traverse(newestCommittedLink, scanByTimestamp, timestamp);
+            rowVersionDataPageReader.traverse(headLink, scanByTimestamp, timestamp);
```

Comment:
Maybe we should rename `scanByTimestamp`

Reply:
I don't follow)

Reply:
We could explicitly mention write intents or something

Reply:
In the name of the class?